### PR TITLE
Multi commit diff summary

### PR DESCRIPTION
**Pull request by tidy-dev**
Built on top of #14682

## Description
This PR introduces a "merged" commit summary view for multi commit selection, and completes the first iteration of multi commit selection. :tada:

I would like this PR to go in as a first draft barring anything that definitely needs fixed (looks like a regression) as this is dev flagged. Then, build on it for any changes. :) Thus, completely open to any feedback (especially if you think from a UI/UX perspective anything could be handled better).

Side note: if you watch the video closely (pay attention to file list), you can see the fun with merge commits.

### Screenshots
https://user-images.githubusercontent.com/75402236/170712834-a9a4a5d4-580f-4b2f-b9a5-5431d55e6d79.mov

## Release notes
Notes: no-notes

### CODE REVIEW

**Comment by tsvetilian-ty** [comment]
@tidy-dev that's a great feature 🎉 

### UI/UX feedback
- When the avatar stack contains more than six people, it overlaps with the text.
image
- If the commit summary description view contains lots of commits, it hides the diff view when expanded. Setting `min-height` to `commit-details` class is potential improvement.

expand-hides-diff

### General feedback
When using the "select all" shortcut, `getPartialBlobContents` return an error. The error occurs in bigger repositories!

shortcut-select-all

The error:
```
Could not load old contents for syntax highlighting
Error: Git returned an unexpected exit code '128' which should be handled by the caller (getPartialBlobContents).
```

**Comment by tidy-dev** [comment]
Thank you for this input @tsvetilian-ty! 

> When the avatar stack contains more than six people, it overlaps with the text.

I knew this was the likely case based on the logic.. I was trying to do minimal change to that component in this pr and kept getting dragged into the rabbit hole..  and was planning to look into making the avatar component handle a the 4+ case more succinctly in a follow up PR.

> commit summary description view contains lots of commits

Great catch!  I will put a max-height on it in this pr. 

> When using the "select all" shortcut, getPartialBlobContents return an error. The error occurs in bigger repositories!

This one is curious; looking at the partial commit logic. I will look into it in another pr as well when delving into the other edge cases I need to resolve. :) 

^ Selecting large groups was edge case I have plans to explore - This is some great stuff!

File: app/src/ui/history/commit-summary.tsx [hunk 1]
```diff
@@ -293,9 +315,20 @@ export class CommitSummary extends React.Component<
...
-    const shortSHA = this.props.commit.shortSha
+  private getShaRef = (useShortSha?: boolean) => {
+    const { commits } = this.props
+    const oldest = useShortSha ? commits[0].shortSha : commits[0].sha
+
+    if (commits.length === 1) {
+      return oldest
+    }
+
+    const latest = useShortSha ? commits.at(-1)?.shortSha : commits.at(-1)?.sha
```
Comment (sergiou87) [inline app/src/ui/history/commit-summary.tsx]: Nit: small refactor here?
```suggestion
    const latestCommit = commits.at(-1)
    const latest = useShortSha ? latestCommit?.shortSha : latestCommit?.sha
```

**Comment by tidy-dev** [comment]
Max height of commit description area set to 400px (still has some diff visible for smallest window size).

image

File: app/src/ui/lib/commit-attribution.tsx [hunk 2]
```diff
@@ -61,24 +61,34 @@ export class CommitAttribution extends React.Component<
...
-      : [author, ...coAuthors]
+      const authors: Array<CommitIdentity | GitAuthor> = committerAttribution
+        ? [author, committer, ...coAuthors]
+        : [author, ...coAuthors]
+
+      for (const a of authors) {
+        if (allAuthors.get(a.toString()) === undefined) {
+          allAuthors.set(a.toString(), a)
+        }
+      }
```
Comment (sergiou87) [inline app/src/ui/lib/commit-attribution.tsx]: I just learned `Set` doesn't get along with objects in JavaScript 😢

File: app/src/ui/lib/commit-attribution.tsx [hunk 3]
```diff
@@ -61,24 +61,34 @@ export class CommitAttribution extends React.Component<
...
 
-    const authors: Array<CommitIdentity | GitAuthor> = committerAttribution
-      ? [author, committer, ...coAuthors]
-      : [author, ...coAuthors]
+      const authors: Array<CommitIdentity | GitAuthor> = committerAttribution
+        ? [author, committer, ...coAuthors]
+        : [author, ...coAuthors]
+
+      for (const a of authors) {
+        if (allAuthors.get(a.toString()) === undefined) {
```
Comment (sergiou87) [inline app/src/ui/lib/commit-attribution.tsx]:
```suggestion
        if (!allAuthors.has(a.toString())) {
```

**Review by sergiou87 — COMMENTED** [review]
Looks good overall! Just a couple of comments 😄

**Comment by tidy-dev** [comment]
Avatar logic appears to only be a little more fiddling to make 4+ authors behave consistently .. seeing as the css was there it just was referencing wrong class  (likely we had not have a scenario that had that many authors to display together before :D) 

https://user-images.githubusercontent.com/75402236/171215099-026a02a4-ca2e-45e9-ab65-c70096308dd1.mov

**Review by sergiou87 — APPROVED** [review]
Thanks for those changes! I tested it with loooong lists of commits (and up to 13 authors) and all worked as expected!

**Review by sergiou87 — APPROVED** [review]
✨

**Review by rpbl3b — COMMENTED** [review]
`app/src/ui/history/commit-summary.tsx`

**Pull request merged**